### ISSUE TITLE
Private DNS Zone Virtual Network Link - Creation fails for cross-tenant

### ISSUE THREAD

**Issue by industrialzombie** [hashicorp/terraform-provider-azurerm#23238]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

0.12.31

### AzureRM Provider Version

3.72.0

### Affected Resource(s)/Data Source(s)

azurerm_private_dns_zone_virtual_network_link

### Terraform Configuration Files

```hcl
terraform {
  required_version = "0.12.31"
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "3.72.0"
    }
  }
}

locals {
  // Fill these in with your own values
  tenant_id_a                    = ""
  tenant_id_b                    = ""
  subscription_id_a              = ""
  virtial_network_id_in_tenant_b = ""
  client_id                      = ""
  client_secret                  = ""
  domain_name                    = ""
}

provider "azurerm" {

  tenant_id       = local.tenant_id_a
  subscription_id = local.subscription_id_a
  client_id       = local.client_id
  client_secret   = local.client_secret

  auxiliary_tenant_ids = [local.tenant_id_b]

  features {
  }
}

resource "azurerm_resource_group" "this" {
  name     = "test_vnet_link_cross_tenant"
  location = "centralus"
}

resource "azurerm_private_dns_zone" "this" {
  name                = local.domain_name
  resource_group_name = azurerm_resource_group.this.name
}

resource "azurerm_private_dns_zone_virtual_network_link" "this" {
  name                  = local.domain_name
  resource_group_name   = azurerm_resource_group.this.name
  private_dns_zone_name = azurerm_private_dns_zone.this.name
  virtual_network_id    = local.virtial_network_id_in_tenant_b
  registration_enabled  = false
}
```

### Debug Output/Panic Output

```shell
azurerm_private_dns_zone_virtual_network_link.this: Creating...

Error: creating/updating Virtual Network Link (Subscription: "[redacted]"
Resource Group Name: "test_vnet_link_cross_tenant"
Private Dns Zone Name: "[redacted]"
Virtual Network Link Name: "[redacted]"): performing CreateOrUpdate: unexpected status 403 with error: LinkedAuthorizationFailed: The client has permission to perform action 'Microsoft.Network/virtualNetworks/join/action' on scope '/subscriptions/[redacted]/resourceGroups/test_vnet_link_cross_tenant/providers/Microsoft.Network/privateDnsZones/[redacted]/virtualNetworkLinks/[redacted]', however the current tenant '[redacted=tenant a]' is not authorized to access linked subscription '[redacted=subscription id in tenant b]'.

  on main.tf line 44, in resource "azurerm_private_dns_zone_virtual_network_link" "this":
  44: resource "azurerm_private_dns_zone_virtual_network_link" "this" {
```

### Expected Behaviour

Private DNS Zone VNet Link should be created successfully.

### Actual Behaviour

Using version 3.66.0, this works as expected.

All versions since 3.67.0 (up to 3.72.0 as of now) give the above error and the VNet link is not created.

### Steps to Reproduce

Arrange:
1. create a subscription in tenant A
2. create a subscription in tenant B
3. create a VNet in subscription B
4. create an app registration with client secret in tenant A
5. grant admin consent to enterprise application for app registration in tenant B

Act:
1. configure locals with details from setup, specifically app registration client secret from step 4
2. terraform apply

### Important Factoids

Admin consent was already granted to the enterprise application and this was working as of version 3.66.0

**Comment by magodo** [comment]
This probably relates to the upgrading of the underlying transport layer to the `github.com/hashicorp/go-azure-sdk` since v3.67.0 (as is stated [here](https://github.com/hashicorp/terraform-provider-azurerm/releases/tag/v3.67.0#:~:text=privatedns%3A%20updating%20to%20use%20the%20base%20layer%20from%20hashicorp/go%2Dazure%2Dsdk%20rather%20than%20Azure/go%2Dautorest)).

But since I didn't find any other complaint about the cross tenant functionality not work on the new transport layer, would you mind to double check your configurations, e.g. retry with the v3.66.0, or with other tools like azure CLI, to make sure it indeed is the provider's bug.

Meanwhile, if you can setup some mitmproxy, you shall be able to see the ARM request, where you can check whether the access tokens for the auxiliary tenant is indeed sent to Azure.

**Comment by industrialzombie** [comment]
Hi @magodo,

Using the az cli I am able to successfully create the resource, so this does appear to be a Terraform issue (or the underlying azure SDK as you mentioned).

I ran the requests through a mitm proxy as suggested.
- The PUT request made by the az cli correctly includes the "x-ms-authorization-auxiliary" header.
- The PUT request made by the Terraform provider **does not** contain this header.

**Comment by magodo** [comment]
Hi @manicminer, do you know whether there is any issue with the cross-tenant scenario when using the `go-azure-sdk` transport layer?

I've checked with the provider setting, and noticed it has correctly specified the auxiliary tenants:

https://github.com/hashicorp/terraform-provider-azurerm/blob/4916dc0c953363e80ec111dd77ab80ec217ce7ac/internal/provider/provider.go#L424-L429

**Comment by industrialzombie** [comment]
Confirmed this is still an issue in 3.73.0 released yesterday.

**Comment by mdanylyuk** [comment]
I have the same problem with 3.68.0 provider version

**Comment by industrialzombie** [comment]
Hey,
I had some more time today to go digging into the code.

If I understand correctly, the change here was to switch the networks service from using the autorest transport to using the new  resource manager client from the go-azure-sdk.

Looking at the codebase, I can see that autorest client gets decorated to call the "AuxiliaryTokens" function of the authorizer. I don't see any equivalent call in the new resource manager client:

It calls the authorizer's "Token" function, but not the "AuxiliaryTokens" function.

https://github.com/hashicorp/go-azure-sdk/blob/main/sdk/client/client.go#L333

https://github.com/hashicorp/go-azure-sdk/blob/main/sdk/auth/autorest/auth.go#L43

This aligns with what I see in the calls being made by the provider in my example; the calls to interact with the resource group are using the autorest client and are correctly have the extra HTTP header, whereas the calls using the new resource manager client do not.

Lastly, I'm pretty new to golang and this is not my codebase so this is all speculation. Please feel free to correct me if I'm way off base here!

**Comment by industrialzombie** [comment]
Also, if I'm right, please can we get the following change reverted until the resource manager client supports auxiliary tokens?

https://github.com/hashicorp/terraform-provider-azurerm/pull/22681/files#diff-71f78de5d72d9452156b0c94b0a4972380457c9cc846db9fc94eb7f1d11009f1

**Comment by manicminer** [comment]
@industrialzombie Thanks for digging into this. Ideally we'd prefer to roll forward and fix the SDK issue rather than roll back to the legacy SDK. I'll look into it and try to issue a fix.

**Comment by manicminer** [comment]
> Hi @manicminer, do you know whether there is any issue with the cross-tenant scenario when using the `go-azure-sdk` transport layer?
> 

@magodo I suspect there may be a bug in the setup or handling of multi-tenant auth, am looking into it.

**Comment by manicminer** [comment]
Thanks again for reporting this authorization issue. We should have a fix out soon.

**Comment by industrialzombie** [comment]
Hey @katbyte

I can see that a test was added for this, but I don't see the underlying issue is resolved yet.
https://github.com/hashicorp/go-azure-sdk/pull/665

If I'm right, please can you re-open this issue?

Also, this has been open now for a while and is still causing problems for my organization.
I appreciate the work that @manicminer has put in on the azure sdk, but is it possible to reconsider the approach of rolling back the regression rather than waiting for rolling forwards?

Thanks

**Comment by manicminer** [comment]
@industrialzombie Sorry about that, I'll reopen this issue. I'm going to try and get the SDK fix in this week to resolve this.

**Comment by manicminer** [comment]
The fix for this was merged in #24063 and this should now be resolved in the next release!
